File: src/main/scala/org/joisen/chapter05/TransformFlatMapTest.scala
package org.joisen.chapter05

import org.apache.flink.api.common.functions.{FilterFunction, FlatMapFunction}
import org.apache.flink.streaming.api.scala._
import org.apache.flink.util.Collector

/**
 * @Author Joisen
 * @Date 2022/11/20 15:52
 * @Version 1.0
 */
object TransformFlatMapTest {
  def main(args: Array[String]): Unit = {

    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
    env.setParallelism(1)

    val stream: DataStream[Event] = env.fromElements(Event("Mary", "./home", 1000L),
      Event("Bob", "./cart", 2000L),Event("Alice", "./cart", 2000L))

    stream.flatMap(new UserFlatMap).print()

    env.execute()
  }

  class UserFlatMap extends FlatMapFunction[Event, String]{
    override def flatMap(t: Event, collector: Collector[String]): Unit = {
      // 如果当前数据是Mary的点击事件，那么直接输出 user
      if(t.user == "Mary"){
        collector.collect(t.user)
      }
      // 如果当前数据是Bob的点击事件，那么就输出user和url
      else if(t.user == "Bob"){
        collector.collect(t.user)
        collector.collect(t.url)
      }

    }
  }

}

File: src/main/java/org/joisen/java/chapter05/TransformFlatMapTest.java
package org.joisen.java.chapter05;

import org.apache.flink.api.common.functions.FilterFunction;
import org.apache.flink.api.common.functions.FlatMapFunction;
import org.apache.flink.api.common.typeinfo.TypeHint;
import org.apache.flink.streaming.api.datastream.DataStreamSource;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.util.Collector;

/**
 * @Author Joisen
 * @Date 2022/12/5 20:48
 * @Version 1.0
 */
public class TransformFlatMapTest {
    public static void main(String[] args) throws Exception {
        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
        env.setParallelism(1);

        DataStreamSource<Event> stream = env.fromElements(new Event("Alice", "./home", 1000L),
                new Event("Mary", "./me", 2000L),
                new Event("Bob", "./cart", 3000L),
                new Event("Cary", "./prod?id=1", 4000L));

//        stream.flatMap(new MyFlatMap()).print();

        stream.flatMap( (Event value, Collector<String> out) -> {
            if(value.user.equals("Bob")) out.collect(value.url);
            else if(value.user.equals("Mary")) {
                out.collect(value.user);
                out.collect(value.url);
                out.collect(value.timestamp.toString());
            }
        } ).returns(new TypeHint<String>() {})
                .print("λ ");

        env.execute();
    }

    public static class MyFlatMap implements FlatMapFunction<Event, String> {

        @Override
        public void flatMap(Event event, Collector<String> collector) throws Exception {
            collector.collect(event.user);
            collector.collect(event.url);
            collector.collect(event.timestamp.toString());
        }
    }

}
